File: backtracking/N&M10.py
import sys
input = sys.stdin.readline

class NandM:
    def __init__(self,n,m,num):
        self.n = n
        self.m = m
        self.num = sorted(num)
        self.arr = [0]*10
        self.result = [0]*m

    def solve(self, k):
        if k==self.m:
            for i in range(self.m):
                print(self.result[i], end=" ")
            print()
            return
        
        st = self.arr[k-1] + 1 if k!=0 else 0
        temp = 0
        for i in range(st,self.n):
            if temp != self.num[i]:
                self.result[k] = self.num[i]
                temp = self.result[k]
                self.arr[k] = i
                self.solve(k+1)

def main():
    n,m = map(int, input().split())
    num = list(map(int, input().split()))
    NM = NandM(n,m,num)
    NM.solve(0)

if __name__ == "__main__":
    main()

File: backtracking/N&M10.cpp
#include<bits/stdc++.h>
using namespace std;

int n,m;
int arr[10];
int num[10];
int result[10];

void solve(int k) {
    if(k==m){
        for(int i=0;i<m;i++) cout<<num[arr[i]]<<" ";
        cout<<"\n";
        return;
    }

    int st = 0;
    if(k!=0) st = arr[k-1]+1;
    int temp = 0;
    for(int i=st;i<n;i++){
        if(temp != num[i]){
            result[k] = num[i];
            arr[k] = i;
            temp = result[k];
            solve(k+1);
        }
    }
}

int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0);

    cin>>n>>m;
    for(int i=0;i<n;i++) cin>>num[i];
    sort(num, num+n);
    solve(0);
}
